modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/GlobalSnapshotAcceptanceManager.scala

```diff
@@ -0,0 +1,145 @@
+import org.tessellation.schema.block.DAGBlock
+import org.tessellation.schema.transaction.{DAGTransaction, RewardTransaction}
+import org.tessellation.sdk.domain.block.processing._
+import org.tessellation.security.signature.Signed
+import org.tessellation.statechannel.{StateChannelOutput, StateChannelSnapshotBinary}
+
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.numeric.NonNegLong
+
+trait GlobalSnapshotAcceptanceManager[F[_]] {
+  def accept(
+    blocksForAcceptance: List[Signed[DAGBlock]],
+    scEvents: List[StateChannelOutput],
+    rewards: SortedSet[RewardTransaction],
+    lastSnapshotContext: GlobalSnapshotInfo,
+    lastActiveTips: SortedSet[ActiveTip],
+    lastDeprecatedTips: SortedSet[DeprecatedTip]
+  ): F[
+    (
+      BlockAcceptanceResult[DAGBlock],
+      SortedMap[Address, NonEmptyList[Signed[StateChannelSnapshotBinary]]],
+      Set[StateChannelOutput],
+      SortedSet[RewardTransaction],
+      GlobalSnapshotInfo
+    )
+  ]
+}
+
+object GlobalSnapshotAcceptanceManager {
+  def make[F[_]: Async: KryoSerializer](
+    blockAcceptanceManager: BlockAcceptanceManager[F, DAGTransaction, DAGBlock],
+    stateChannelEventsProcessor: GlobalSnapshotStateChannelEventsProcessor[F],
+    collateral: Amount
+  ) = new GlobalSnapshotAcceptanceManager[F] {
+
+    def accept(
+      blocksForAcceptance: List[Signed[DAGBlock]],
+      scEvents: List[StateChannelOutput],
+      rewards: SortedSet[RewardTransaction],
+      lastSnapshotContext: GlobalSnapshotInfo,
+      lastActiveTips: SortedSet[ActiveTip],
+      lastDeprecatedTips: SortedSet[DeprecatedTip]
+    ) = for {
+      acceptanceResult <- acceptBlocks(blocksForAcceptance, lastSnapshotContext, lastActiveTips, lastDeprecatedTips)
+
+      (scSnapshots, currencySnapshots, returnedSCEvents) <- stateChannelEventsProcessor.process(lastSnapshotContext, scEvents)
+      sCSnapshotHashes <- scSnapshots.toList.traverse { case (address, nel) => nel.head.hashF.map(address -> _) }
+        .map(_.toMap)
+      updatedLastStateChannelSnapshotHashes = lastSnapshotContext.lastStateChannelSnapshotHashes ++ sCSnapshotHashes
+      updatedLastCurrencySnapshots = lastSnapshotContext.lastCurrencySnapshots ++ currencySnapshots
+
+      transactionsRefs = lastSnapshotContext.lastTxRefs ++ acceptanceResult.contextUpdate.lastTxRefs
+
+      (updatedBalancesByRewards, acceptedRewardTxs) = acceptRewardTxs(
+        lastSnapshotContext.balances ++ acceptanceResult.contextUpdate.balances,
+        rewards
+      )
+
+      maybeMerkleTree <- updatedLastCurrencySnapshots.merkleTree[F]
+      updatedLastCurrencySnapshotProofs <- maybeMerkleTree.traverse { merkleTree =>
+        updatedLastCurrencySnapshots.toList.traverse {
+          case (address, state) =>
+            (address, state).hashF
+              .map(merkleTree.findPath(_))
+              .map(_.get) // NOTE: path must exists, otherwise it's an unexpected state and should throw
```

Comment:
Maybe change the model so we don't need to throw? A tuple or simple product class would fix the issue.

Types used in `GlobalSnapshot` were designed to not allow representing illegal states (in example `BlockAsActiveTip`). It would be nice to continue with this semantics.

Reply:
how exactly you would fix that? we have many places in the codebase that we assume the state and throws unexpected state. `findPath` returns an `Option` and there is no way to throw from `get` other than programmatic error

Reply:
Right, I thought it was a simple lookup in the map. Then please throw the error that says that the Merkle tree is invalid. NPE in the logs isn't too informative, especially since the stack traces aren't very informative currently.

Reply:
changing :+1: